framework/packages/abstract-std/src/native/ica_client.rs

```diff
@@ -0,0 +1,173 @@
+use crate::objects::TruncatedChainId;
+use cosmwasm_schema::QueryResponses;
+use cosmwasm_std::CosmosMsg;
+use cw_ownable::{cw_ownable_execute, cw_ownable_query};
+
+pub use action::{IcaAction, IcaActionResponse, IcaExecute};
+pub use chain_type::{CastChainType, ChainType};
+
+pub use polytone_evm::EVM_NOTE_ID;
+pub use polytone_evm::POLYTONE_EVM_VERSION;
+
+/// This needs no info. Owner of the contract is whoever signed the InstantiateMsg.
+#[cosmwasm_schema::cw_serde]
+pub struct InstantiateMsg {
+    pub ans_host_address: String,
+    pub version_control_address: String,
+}
+
+#[cosmwasm_schema::cw_serde]
+pub struct MigrateMsg {}
+
+#[cw_ownable_execute]
+#[cosmwasm_schema::cw_serde]
+#[derive(cw_orch::ExecuteFns)]
+pub enum ExecuteMsg {}
+
+#[cw_ownable_query]
+#[cosmwasm_schema::cw_serde]
+#[derive(QueryResponses, cw_orch::QueryFns)]
+pub enum QueryMsg {
+    /// Returns config
+    /// Returns [`ConfigResponse`]
+    #[returns(ConfigResponse)]
+    Config {},
+
+    #[returns(IcaActionResult)]
+    IcaAction {
+        // Proxy address used to query polytone implementations or proxy itself.
+        proxy_address: String,
+        // Chain to send to
+        chain: TruncatedChainId,
+        // Queries go first
+        actions: Vec<IcaAction>,
+    },
+}
+
+#[cosmwasm_schema::cw_serde]
+pub struct ConfigResponse {
+    pub ans_host: String,
+    pub version_control_address: String,
+}
+
+#[cosmwasm_schema::cw_serde]
+pub struct IcaActionResult {
+    /// messages that call the underlying implementations (be it polytone/cw-ica-controller/etc)
+    pub msgs: Vec<CosmosMsg>,
+}
+
+mod chain_type {
+    use std::fmt::Display;
+
+    use crate::objects::TruncatedChainId;
+    use crate::registry::*;
+
+    #[derive(Debug, Clone, Copy, PartialEq)]
+    pub enum ChainType {
+        Evm,
+        Cosmos,
+    }
+
+    impl Display for ChainType {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            match self {
+                ChainType::Evm => write!(f, "EVM"),
+                ChainType::Cosmos => write!(f, "Cosmos"),
+            }
+        }
+    }
+
+    pub trait CastChainType {
+        fn chain_type(&self) -> Option<ChainType>;
+    }
+
+    impl CastChainType for TruncatedChainId {
+        // Return the type of chain based on the chain-id.
+        // Note: chain-ids for EVM chains are numbers!
+        fn chain_type(&self) -> Option<ChainType> {
+            // TODO: Should be a better way to have lookup table, maybe this will be nice:
+            // https://crates.io/crates/phf
+            let chains = map_macro::hash_map! {
```

Comment:
Have you considered lookup table alternative? What do you think about phf?

Reply:
This is not possible because hf doesn't support keys of the type ARCHWAY[0] it seems here